x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java

```diff
@@ -480,8 +481,15 @@ private void authorizeAction(
                 l.onResponse(result);
             }));
         } else if (isIndexAction(action)) {
-            final ProjectMetadata projectMetadata = projectResolver.getProjectMetadata(clusterService.state());
-            assert projectMetadata != null;
+            // We use a supplier here to avoid resolving the project in cases where it's not going to be used
+            // The most common case is when a child action is authorized based on its parent, in that case we never need
+            // to lookup indices, so we never need the project metadata.
+            // But we do want to optimize by only looking up the project once for each action, so we use a cached supplier.
+            final Supplier<ProjectMetadata> projectMetadataSupplier = CachedSupplier.wrap(() -> {
```

Comment:
This adds a volatile read which may not help with performance for single project case. I wonder whether we could conditionally return a simple supplier when there is only a single project, e.g.:

```java
if (clusterService.state().metadata().projects().size() == 1) {
    final var projectMetadata = projectResolver.getProjectMetadata(clusterService.state());
    projectMetadataSupplier = () -> projectMetadata;
} else {
    ...
}
```

I don't particularly like checking number of projects outside of Metadata. But I don't have a great suggestion either. Plus it still has to create a lambda for the supplier.